### PR TITLE
DavidWidjaja-dev

### PR DESCRIPTION
**Pull request by DavidWidjaja-137**
Speed Controller node

### CODE REVIEW

File: Speed_Controller/src/Encoder/encoder.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	
+	TIM1->CR1 |= 0x1;	
+	
+}
+
+/**
+ * Reads a value from the encoder
+ */
+uint16_t EncoderRead (void) {
+
```
Comment (cng123) [inline Speed_Controller/src/Encoder/encoder.c:44]: Should we move the filter code here?

File: Speed_Controller/src/Timer/Timer.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	TIM2->CR1 &= (0x11UL << 8);	//set the timer to use default clock division
+
+	TIM2->CR1 |= 0x1UL << 3;	//set the timer to use one pulse mode
+	
+	TIM2->CR1 |= 0x1UL;			//enable TIM2
+	
+	TIM2->DIER |= 0x1UL;		//enable timing interrupt
+	
+	NVIC_EnableIRQ(TIM2_IRQn);	//setup interrupt handler
+
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.c]: extra space

File: Speed_Controller/src/Timer/Timer.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	TIM2->CNT = 0x1;
+	TIM2->CR1 |= 0x1UL;			//re-enable TIM2
+
+}
+
+/**
+ * Stops the counter on the timer
+ */
+void StopTimer(void){
+
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.c]: spacing

File: Speed_Controller/src/Timer/Timer.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#include "stm32f10x.h"
+
+#define TIM2_PRESCALER 7199
+
+#define TRUE 1
+#define FALSE 0
+
+/** Interrupt handler for Timer 2.
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.h:7]: formatting

File: Speed_Controller/src/Timer/Timer.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+#define TRUE 1
+#define FALSE 0
+
+/** Interrupt handler for Timer 2.
+ */
+void TIM2_IRQHandler(void);
+
+
+/** Initializes a timer to trigger a timing interrupt periodically
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.h:11]: formatting

File: Speed_Controller/src/Timer/Timer.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#include "stm32f10x.h"
+
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.h:2]: should probably have header guards for this file

File: Speed_Controller/src/main.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+#include "CAN.h"
+#include "virtual_com.h"
+#include "Timer/Timer.h"
+
+#define DRIVE_CONTROL_ID 0x500
+
+union {
+	float float_var;
+	uint8_t chars[4];
+} u;
```
Comment (cng123) [inline Speed_Controller/src/main.c:14]: we should probably use better names

File: Speed_Controller/src/main.c [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+union {
+	float float_var;
+	uint8_t chars[4];
+} v;
```
Comment (cng123) [inline Speed_Controller/src/main.c:19]: same as above

File: Speed_Controller/src/main.c [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	CAN_drive.data[5] = 0;
+	CAN_drive.data[6] = 0;
+	CAN_drive.data[7] = 0;
+	
+	CAN_regen.data[4] = 0;
+	CAN_regen.data[5] = 0;
+	CAN_regen.data[6] = 0;
+	CAN_regen.data[7] = 0;
+		
+	RCC->APB2ENR |= 0x1 << 4;
```
Comment (cng123) [inline Speed_Controller/src/main.c:62]: should this be in its own module?

File: Speed_Controller/src/main.c [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	volatile uint16_t ADC_reading;
+	uint16_t old_ADC_reading = 0x000;
+	uint16_t old_encoder_reading = 0x000;
+	uint8_t regen_enabled;
+	uint8_t current_direction;
+	uint8_t previous_direction = 0;
+	
+	while(1) 
+	{
+
```
Comment (cng123) [inline Speed_Controller/src/main.c:76]: spacing

File: Speed_Controller/src/main.c [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+			CAN_drive.data[2] = v.chars[2];
+			CAN_drive.data[3] = v.chars[3];			
+			
+			previous_direction = current_direction;
+		}
+		
+		//If regen is enabled AND ADC count changed, send new regen CAN message and restart timer
+		if( regen_enabled && (old_ADC_reading != ADC_reading) )
+		{
+
```
Comment (cng123) [inline Speed_Controller/src/main.c:98]: spacing

File: Speed_Controller/src/main.c [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+			RestartTimer();
+		}		
+		//if encoder count changed, send new drive CAN message and restart timer
+		else if(old_encoder_reading != encoder_reading && (ADC_reading == 0 || !regen_enabled))
+		{
+			
+			u.float_var = (float)( (float) encoder_reading/PEDAL_MAX);
+			
+			//Any overflow should be treated as zero current to the motor
+			if (u.float_var > 1.0 || u.float_var < 0.0)
```
Comment (cng123) [inline Speed_Controller/src/main.c]: should this be moved to the encoderRead function?

File: Speed_Controller/src/main.c [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+			}
+			//If the driver is not braking, send the previous CAN drive message.
+			else
+			{	
+				CANSend(&CAN_drive);
+			}
+			
+			timeoutFlag = FALSE;
+			
+			old_ADC_reading = ADC_reading;
```
Comment (cng123) [inline Speed_Controller/src/main.c:153]: why is this in each branch? why can't they just be outside the cases and at the end of the while loop?

File: Peripherals/CAN/CAN.c [hunk 14]
```diff
@@ -22,7 +22,7 @@
 	 
 	// Set bit rates 
 	CAN1->BTR &= ~(((0x03) << 24) | ((0x07) << 20) | ((0x0F) << 16) | (0x1FF)); 
-	CAN1->BTR |=  (((2-1) & 0x07) << 20) | (((9-1) & 0x0F) << 16) | ((6-1) & 0x1FF);
+	CAN1->BTR |=  (((2-1) & 0x07) << 20) | (((15-1) & 0x0F) << 16) | ((2-1) & 0x1FF);
```
Comment (cng123) [inline Peripherals/CAN/CAN.c:25]: we should probably do something about this and add support to be able to change this via defining a constant

File: Speed_Controller/src/ADC/ADC.c [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	
+	// Configure ADC2
+	// Enable ADC clock
+	RCC->APB2ENR |= RCC_APB2ENR_ADC2EN;
+	
+	// Enable config registers
+	// By default, CR1 is set correctly
+	ADC2->CR2 |= ADC_CR2_CONT;
+	
+	// TODO: check sample time to see diff
```
Comment (cng123) [inline Speed_Controller/src/ADC/ADC.c]: we can remove this, don't remember there was a difference

Comment (cng123) [reply Speed_Controller/src/ADC/ADC.c]: remove the comment, not the code

File: Speed_Controller/src/Timer/Timer.c [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+/** Initializes a timer to trigger a timing interrupt periodically
+ * @param period: period for triggering the timing interrupt
+ */
+void TimerInit(int period){
+	
+	RCC->APB1ENR |= 0x1;		//enable TIM2
+	
+	TIM2->PSC = TIM2_PRESCALER;	//Processor speed is ~72MHZ, so set prescaler value for scaling,
+								//Clock frequency of 10kHz
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.c]: spacing

File: Speed_Controller/src/Timer/Timer.h [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+#define TRUE 1
+#define FALSE 0
+
+/** Interrupt handler for Timer 2.
+ */
+void TIM2_IRQHandler(void);
+
+
+/** Initializes a timer to trigger a timing interrupt periodically
+ * @param period: period for triggering the timing interrupt
```
Comment (cng123) [inline Speed_Controller/src/Timer/Timer.h]: units?

File: Speed_Controller/src/main.c [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+union {
+	float float_var;
+	uint8_t chars[4];
+} v;
+
+int main(void){
+	CANInit();
+	EncoderInit();
+	ADCInit();
+	TimerInit(100);
```
Comment (cng123) [inline Speed_Controller/src/main.c:25]: we should probably try increasing the period until the motor controller stops operating smoothly

**Review by cng123 — CHANGES_REQUESTED** [review]

File: Speed_Controller/src/Encoder/encoder.c (hunk 1 above)
Comment (DavidWidjaja-137) [reply Speed_Controller/src/Encoder/encoder.c:44]: yea probably

**Comment by DavidWidjaja-137** [comment]
changes made to speed controller

**Pull request closed without merging**